enterprise/tailnet/pgcoord.go

```diff
@@ -0,0 +1,1207 @@
+	eb := backoff.NewExponentialBackOff()
+	eb.MaxElapsedTime = 0 // retry indefinitely
+	eb.MaxInterval = dbMaxBackoff
+	bkoff := backoff.WithContext(eb, q.ctx)
+	for {
+		mk, err := q.workQ.acquire()
+		if err != nil {
+			// context expired
+			return
+		}
+		err = backoff.Retry(func() error {
+			return q.query(mk)
+		}, bkoff)
+		if err != nil {
+			bkoff.Reset()
+		}
+		q.workQ.done(mk)
+	}
+}
+
+func (q *querier) query(mk mKey) error {
+	var mappings []mapping
+	var err error
+	if mk.clientsOfAgent {
+		mappings, err = q.queryClientsOfAgent(mk.agent)
+		if err != nil {
+			return err
+		}
+	} else {
+		mappings, err = q.queryAgent(mk.agent)
+		if err != nil {
+			return err
+		}
+	}
+	q.mu.Lock()
+	mpr, ok := q.mappers[mk]
+	q.mu.Unlock()
+	if !ok {
+		q.logger.Debug(q.ctx, "query for missing mapper",
+			slog.F("agent_id", mk.agent), slog.F("clients_of_agent", mk.clientsOfAgent))
+		return nil
+	}
+	mpr.mappings <- mappings
+	return nil
+}
+
+func (q *querier) queryClientsOfAgent(agent uuid.UUID) ([]mapping, error) {
+	clients, err := q.store.GetTailnetClientsForAgent(q.ctx, agent)
+	if err != nil && !xerrors.Is(err, sql.ErrNoRows) {
+		return nil, err
+	}
+	mappings := make([]mapping, 0, len(clients))
+	for _, client := range clients {
+		node := new(agpl.Node)
+		err := json.Unmarshal(client.Node, node)
+		if err != nil {
+			q.logger.Error(q.ctx, "failed to unmarshal node", slog.Error(err))
+			return nil, backoff.Permanent(err)
+		}
+		mappings = append(mappings, mapping{
+			client:      client.ID,
+			agent:       client.AgentID,
+			coordinator: client.CoordinatorID,
+			updatedAt:   client.UpdatedAt,
+			node:        node,
+		})
+	}
+	return mappings, nil
+}
+
+func (q *querier) queryAgent(agentID uuid.UUID) ([]mapping, error) {
+	agents, err := q.store.GetTailnetAgents(q.ctx, agentID)
+	if err != nil && !xerrors.Is(err, sql.ErrNoRows) {
+		return nil, err
+	}
+	mappings := make([]mapping, 0, len(agents))
+	for _, agent := range agents {
+		node := new(agpl.Node)
+		err := json.Unmarshal(agent.Node, node)
+		if err != nil {
+			q.logger.Error(q.ctx, "failed to unmarshal node", slog.Error(err))
+			return nil, backoff.Permanent(err)
+		}
+		mappings = append(mappings, mapping{
+			agent:       agent.ID,
+			coordinator: agent.CoordinatorID,
+			updatedAt:   agent.UpdatedAt,
+			node:        node,
+		})
+	}
+	return mappings, nil
+}
+
+func (q *querier) subscribe() {
+	eb := backoff.NewExponentialBackOff()
+	eb.MaxElapsedTime = 0 // retry indefinitely
+	eb.MaxInterval = dbMaxBackoff
+	bkoff := backoff.WithContext(eb, q.ctx)
+	var cancelClient context.CancelFunc
+	err := backoff.Retry(func() error {
+		cancelFn, err := q.pubsub.SubscribeWithErr(eventClientUpdate, q.listenClient)
+		if err != nil {
+			q.logger.Warn(q.ctx, "failed to subscribe to client updates", slog.Error(err))
+			return err
+		}
+		cancelClient = cancelFn
+		return nil
+	}, bkoff)
+	if err != nil {
+		// this should only happen if context is canceled
```

Comment:
Should we add a sanity check to verify that this is a context related error? Otherwise we can log `BUG: ...`